### PR TITLE
Metrics with OpenTelemetry

### PR DESCRIPTION
**Pull request by javipacheco**
This PR is the first approach for adding metrics to Xef. 

The new `Metric` component is added to `Conversation` in the same way that we are adding `VectorStore`. The metrics contain 2 implementations:

- `LogsMetric`: it is the default implementation in `core`. This implementation adds logs in the console
- `OpenTelemetryMetric`: it is an integration with OpenTelemetry

Here is an image using `JaegerUI`

Screenshot 2023-10-06 at 13 27 31

You have comments in the `Animal.kt` example for testing OpenTelemetry locally

### CODE REVIEW

**Comment by raulraja** [comment]
this is looking great!, thank you @javipacheco !

File: integrations/opentelemetry/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    implementation(libs.opentelemetry.extension.kotlin)
+    implementation(libs.opentelemetry.exporter.otlp)
+
+    testImplementation(libs.junit.jupiter.api)
+    testImplementation(libs.kotest.property)
+    testImplementation(libs.kotest.framework)
+    testImplementation(libs.kotest.assertions)
+    testRuntimeOnly(libs.kotest.junit5)
+}
+
```
Comment (Montagon) [inline integrations/opentelemetry/build.gradle.kts:41]: I think you should add the spotless plugin here. Something like:

Inside `plugins`:
`alias(libs.plugins.spotless)`

And outside:

```gradle
spotless {
  kotlin {
    target("**/*.kt")
    ktfmt().googleStyle().configure {
      it.setRemoveUnusedImport(true)
    }
  }
}
```

File: integrations/opentelemetry/src/main/kotlin/com/xebia/functional/xef/opentelemetry/OpenTelemetryConfig.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+data class OpenTelemetryConfig(
+    val endpointConfig: String,
+    val defaultScopeName: String,
+    val serviceName: String
+) {
+
+    fun newInstance(): OpenTelemetry {
+        val jaegerOtlpExporter: OtlpGrpcSpanExporter = OtlpGrpcSpanExporter.builder()
+            .setEndpoint(endpointConfig)
+            .setTimeout(30, TimeUnit.SECONDS)
```
Comment (Montagon) [inline integrations/opentelemetry/src/main/kotlin/com/xebia/functional/xef/opentelemetry/OpenTelemetryConfig.kt:22]: Should be this parameter taken from config or environment?

**Review by Montagon — APPROVED** [review]
Just some comments. Great job, @javipacheco! 🚀

File: integrations/opentelemetry/src/main/kotlin/com/xebia/functional/xef/opentelemetry/OpenTelemetryConfig.kt (hunk 2 above)
Comment (javipacheco) [reply integrations/opentelemetry/src/main/kotlin/com/xebia/functional/xef/opentelemetry/OpenTelemetryConfig.kt:22]: I think that it's not necessary for now

File: integrations/opentelemetry/build.gradle.kts (hunk 1 above)
Comment (javipacheco) [reply integrations/opentelemetry/build.gradle.kts:41]: We should add spotless in a new PR

**Review by Montagon — APPROVED** [review]

**Pull request merged**